File: 0122-best-time-to-buy-and-sell-stock-ii/0122-best-time-to-buy-and-sell-stock-ii.java
class Solution {
    public int maxProfit(int[] prices) {
        int n = prices.length;
        int[] dp = new int[n];
        
        dp[0] = 0;
        
        for (int i = 1; i < n; i++) {
            for (int j = 0; j < i; j++) {
                if (prices[j] >= prices[i]) {
                    dp[i] = Math.max(dp[i], dp[j]);
                }
                else {
                    dp[i] = Math.max(dp[i], dp[j] + (prices[i] - prices[j]));
                }
            }
        }
        
        return dp[n - 1];
    }
}

File: 0122-best-time-to-buy-and-sell-stock-ii/0122-best-time-to-buy-and-sell-stock-ii.cpp
class Solution {
public:
    int maxProfit(vector<int>& prices) {
        int n = prices.size();
        std::vector<int> dp(n);
        
        for (int i = 1; i < n; i++) {
            if (prices[i - 1] >= prices[i]) {
                dp[i] = dp[i - 1];
            }
            else {
                dp[i] = dp[i - 1] + (prices[i] - prices[i - 1]);
            }
        }
        
        return dp[n - 1];
    }
};